### ISSUE TITLE
Allow weight_function wherever edge weight is used

### ISSUE THREAD

**Issue by dschult** [networkx/networkx#4085]
We have a nice treatment of edge weights in the shortest_path sub-package: the weight can be the name of the edge attribute that holds the weight, but it can also be a function of the data dict for an edge. Thus the weight can be static, or it can be computed as needed.

There are a number of other places in the code-base where edge weights are used. It would be good to enable the weight function interface throughout.

**Comment by dschult** [comment]
I'll try to construct a list of functions where a "weight" argument could be upgraded to the ```_weight_function``` approach.  This might be a good topic for a virtual sprint (lots of places with similar code-fixes to implement.

**Comment by adnanmuttaleb** [comment]
@dschult I tried to fix the `reaching.py`  as you suggested, but in order to make it completely consistent you have to call `_weight_function`  in each function that use the `weight` parameter directly in the invocation tree. Or maybe we can assume that `weight` is always a callable of type `Callable(u: Node, v :Node, data:Mapping)` in all functions. To do so we may have something like:

```
from functools import wraps

def _weight_function(G, weight):
    if callable(weight):
        return weight
    if G.is_multigraph():
        return lambda u, v, data: min(attr.get(weight, 1) for attr in data.values())
    return lambda u, v, data: data.get(weight, 1)


def weighted(func):
  @wraps(func)
  def wrap(G, *args, weight='weight', **kwargs):
    weight = _weight_function(G, weight)
    return func(G, *args, weight=weight, **kwargs)
```

usage:

```
@weighted
def single_source_dijkstra_path(G, source, cutoff=None, weight='weight'):
  pass

@weighted
def global_reaching_centrality(G, weight=None, normalized=True):
  pass
```

**Comment by dschult** [comment]
Yes, I think the decorator approach is probably best.  It'd be great if it could work for other weight-like parameters like "cost" or "capacity", but that might make it too complex.

We should move ```_weight_function``` and its decorator into the networkx/utils directory, perhaps in the decorators.py module.

**Comment by adnanmuttaleb** [comment]
Great, I will work on it. We may create a general parameterized decorator that work for any keyword argument `weight`, `capacity`..etc.  I will submit PR soon hopefully.

**Comment by adnanmuttaleb** [comment]
After doing this initial draft a few question raised:

1. what is the meaning of `is_weighted` if weight now is always a function.
2. what is the expected behavior if client passed `weight` as None, (consider the graph as not-weighted or something else).
3. I have included the graph object in the `wrap_attr` function. `wrap_attr`  is same as `_weight_function` but more generic. I assumed that maybe latter there will be use cases that need the graph to compute the property (e.g weight), or I am wrong?

**Comment by dschult** [comment]
1. I think we should still allow ```is_weighted``` to verify that every edge has the given attribute key. That is a useful function. Remember, weight is NOT always a function. The user sees it as usually a string.  The developer can think of it as a function, but the user does not.

2. When does it occur that weight=None?  I believe this should only happen for functions that must parse the weight parameter to determine if an unweighted algorithm should be used. In that case, weight==None (I think this is true everywhere it is used) means  use an unweighted algorithm.

BUT...  I also think this function doesn't need to use this decorator. It should process the weight parameter appropriately and then call the weighted algorithm (or not). The weighted algorithm function should have the decorator.  We should NOT have code that calls the decorator code more than once. IMO.

3. I don't think we should include the graph parameter in the weight function. I have never seen a use case for that and if it is needed, it will be the same for all edges. So the user can include that in the namespace accessible to their weight function. This function will get called a lot... so we want it to be light and fast.

4. It is true that ```computed_attrs``` is idempotent. But we don't want to run code just to run code. I feel like the decorator should only be used on functions that actually use the weight parameter -- not functions that call functions that use the weight parameter. Does that make sense?

Thanks!

**Comment by adnanmuttaleb** [comment]
Thank you for your response. All points are clear to me now, I will update ASAP.

**Comment by adnanmuttaleb** [comment]
@dschult I made an update: removed G parameter, applied `computed_attr` decorator only on functions that actually use the weight parameter. 

But still I have a question: There are some functions that test the `weight` parameter for nullity to choose the appropriate algorithm, while other functions assume that `weight` should be a function that returns 1 if `weight` is passed as `None` (by calling `_weight_function` inside the function body), now what should I return from `wrap_attr` in case of a `None`? return `None` or return a function that always evaluate to constant value? in both cases there will be some test cases that will fail.

**Comment by dschult** [comment]
We need to fix the functions that send ```None``` into ```_weight_function``` expecting it to give a function that returns 1 always.  They should be smarter than that.  I suppose we could try to think of a way to make ```wrap_attr``` decorator return a function that quickly returns 1. But I'd have to look at the functions that handle ```None``` that way.  I hope there aren't many.

I see your edited resolution and I think it is a reasonable approach. The decorator could return ```None``` if the input is ```None```. If there is no default specified for the keyword (does this happen?  I hope not often), then maybe we should raise an exception.

Other thoughts:  (bikeshedding perhaps) Can we change names so that it is clear this is for edge attributes instead of any attributes?  For example:  ```wrap_edge_attribute```.  Also, these aren't all computed attributes. We just want people to be allowed to use computed attributes.  Perhaps ```@edge_attribute``` would be a better name?  And finally, would the API be simpler and just as good if we restricted it to one attribute?  But could be still handle more than one attribute by adding more than one decorator?
```python
@edge_attribute("weight")
@edge_attribute("cost")
@edge_attribute("capacity")
def some_flow_function(G, cost="cost", "capacity"="capacity", weight="weight"):
```

Thoughts?

**Comment by dschult** [comment]
In #4976 we have a case where the `weight` function returned should do more than return the weight value. In the case of multigraph betweenness we want (for a node-pair) to return the minimum weight **and** a list of edge keys that attain that weight.  But the overall goal of this decorator is the same: let the user write their own weight function, or input a string.  So it appears we want the decorator to allow flexible default treatment of the weight attribute if a string is specified.  For the example mentioned above we only need to change the default for multi graphs, but I can foresee a need for a default for graphs and a different default for multigraphs.   Perhaps we can use the following decorator signature:

`@edge_attribute(parameter_name, handle_string=get_default_1, handle_string_multigraph=get_min_with_default_1)`

usage is normally:  `@edge_attribute(“weight”)`
But for `edge_betweenness_centrality` we would define a function to handle multigraphs and return the min weight and a list of keys and then pass that into the `edge_attribute` decorator:
```python
def get_min_weight_and_keys(weight, u, v, keydict):
 ```
